File: leetcode/dota2_senate/python/solution.py
from collections import deque


class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        """
        Time complexity: O(n)
        Auxiliary space complexity: O(n)
        Tags:
            DS: queue
            A: greedy
        """
        votes = deque(senate)
        r = 0
        
        while (
            len(votes) > abs(r)
        ):
            vote = votes.popleft()
            
            if vote == "R":
                if r >= 0:
                    votes.append("R")
                r += 1

            else:
                if r <= 0:
                    votes.append("D")
                r -= 1

        return "Radiant" if votes[0] == "R" else "Dire"
        

class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        """
        Time complexity: O(n)
        Auxiliary space complexity: O(n)
        Tags:
            DS: queue
            A: greedy
        """
        # (radiant index)
        radiant_votes = deque()
        # (dire index)
        dire_votes = deque()

        for index, vote in enumerate(senate):
            if vote == "R":
                radiant_votes.append(index)
            else:
                dire_votes.append(index)
            
        while radiant_votes and dire_votes:
            if radiant_votes[0] < dire_votes[0]:
                dire_votes.popleft()
                radiant_votes.append(radiant_votes.popleft() + len(senate))
            else:
                radiant_votes.popleft()
                dire_votes.append(dire_votes.popleft() + len(senate))
            
        return "Radiant" if radiant_votes else "Dire"


print(Solution().predictPartyVictory("RD") == "Radiant")
print(Solution().predictPartyVictory("DR") == "Dire")
print(Solution().predictPartyVictory("RDD") == "Dire")
print(Solution().predictPartyVictory("RDDR") == "Radiant")
print(Solution().predictPartyVictory("DRRD") == "Dire")
print(Solution().predictPartyVictory("DDDDRRDDDRDRDRRDDRDDDRDRRRRDRRRRRDRDDRDDRRDDRRRDDRRRDDDDRRRRRRRDDRRRDDRDDDRRRDRDDRDDDRRDRRDRRRDRDRDR") == "Dire")

File: leetcode/dota2_senate/javascript/solution.js
import { Queue } from '@datastructures-js/queue'


class Solution {
   /**
    * Time complexity: O(n)
    * Auxiliary space complexity: O(n)
    * Tags:
    *     DS: queue
    *     A: greedy
    * @param {string} senate
    * @return {string}
    */
   predictPartyVictory(senate) {
      const votes = new Queue(Array.from(senate));
      let r = 0;

      while (votes.size() > Math.abs(r)) {
         const vote = votes.dequeue();

         if (vote === 'R') {
            if (r >= 0)
               votes.enqueue('R');
            r++;

         } else {
            if (r <= 0)
               votes.enqueue('D');
            r--;
         }
      }

      return votes.front() == "R" ? 'Radiant' : 'Dire';
   };

   /**
    * Time complexity: O(n)
    * Auxiliary space complexity: O(n)
    * Tags:
    *     DS: queue
    *     A: greedy
    * @param {string} senate
    * @return {string}
    */
   predictPartyVictory(senate) {
      const radiantVotes = new Queue();
      const direVotes = new Queue();

      for (let index = 0; index < senate.length; index++) {
         const senator = senate[index];
         if (senator === 'R')
            radiantVotes.enqueue(index);
         else
            direVotes.enqueue(index);
      }

      while (
         radiantVotes.size() &&
         direVotes.size()
      ) {
         if (radiantVotes.front() < direVotes.front()) {
            radiantVotes.enqueue(radiantVotes.front() + senate.length);
         } else {
            direVotes.enqueue(direVotes.front() + senate.length);
         }
         radiantVotes.dequeue();
         direVotes.dequeue()
      }

      return radiantVotes.isEmpty() ? 'Dire' : 'Radiant'
   };
}


const predictPartyVictory = new Solution().predictPartyVictory;
console.log(new Solution().predictPartyVictory('RD') === 'Radiant')
console.log(new Solution().predictPartyVictory('DR') === 'Dire')
console.log(new Solution().predictPartyVictory('RDD') === 'Dire')
console.log(new Solution().predictPartyVictory('RDDDRR') === 'Dire')
console.log(new Solution().predictPartyVictory('RDDR') === 'Radiant')
console.log(new Solution().predictPartyVictory('DRRD') === 'Dire')
console.log(new Solution().predictPartyVictory('DDDDRRDDDRDRDRRDDRDDDRDRRRRDRRRRRDRDDRDDRRDDRRRDDRRRDDDDRRRRRRRDDRRRDDRDDDRRRDRDDRDDDRRDRRDRRRDRDRDR') === 'Dire')
